File: platforms/android/samples/CheckoutKitAndroidDemo/app/src/main/java/com/shopify/checkout_kit_android_demo/cart/CartBootstrapLink.kt
package com.shopify.checkout_kit_android_demo.cart

import android.net.Uri
import com.shopify.checkout_kit_android_demo.common.ID

data class CartBootstrapLink(
    val variantId: ID?,
    val productIndex: Int?,
    val quantity: Int,
) {
    companion object {
        private const val SCHEME = "checkout-kit-android"
        private const val HOST = "cart"

        fun parse(uri: Uri): CartBootstrapLink? {
            if (uri.scheme != SCHEME) return null

            if (uri.host != HOST) {
                throw CartBootstrapException("Unsupported cart bootstrap path")
            }

            val variantId = uri.getQueryParameter("variantId")?.trim()
            val productIndexParam = uri.getQueryParameter("productIndex")?.trim()
            val quantityParam = uri.getQueryParameter("quantity")?.trim() ?: "1"
            val quantity = quantityParam.toIntOrNull()

            if (quantity == null || quantity < 1) {
                throw CartBootstrapException("quantity must be a positive integer")
            }

            if (!variantId.isNullOrEmpty() && !productIndexParam.isNullOrEmpty()) {
                throw CartBootstrapException("Use variantId or productIndex, not both")
            }

            if (!variantId.isNullOrEmpty()) {
                return CartBootstrapLink(variantId = ID(variantId), productIndex = null, quantity = quantity)
            }

            if (productIndexParam.isNullOrEmpty()) {
                throw CartBootstrapException("Missing variantId or productIndex")
            }

            val productIndex = productIndexParam.toIntOrNull()
            if (productIndex == null || productIndex < 0) {
                throw CartBootstrapException("productIndex must be a non-negative integer")
            }

            return CartBootstrapLink(variantId = null, productIndex = productIndex, quantity = quantity)
        }
    }
}

class CartBootstrapException(message: String) : IllegalArgumentException(message)

File: platforms/swift/Samples/CheckoutKitSwiftDemo/CheckoutKitSwiftDemo/Sources/App/CartBootstrapLink.swift
import Foundation

struct CartBootstrapLink {
    let variantId: String?
    let productIndex: Int?
    let quantity: Int

    private static let scheme = "checkout-kit-swift"
    private static let host = "cart"

    static func parse(_ url: URL) throws -> CartBootstrapLink? {
        guard url.scheme == scheme else { return nil }

        guard url.host == host else {
            throw CartBootstrapError.unsupportedPath
        }

        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        let queryItems = components?.queryItems ?? []

        func queryValue(_ name: String) -> String? {
            return queryItems
                .first(where: { $0.name == name })?
                .value?
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }

        let variantId = queryValue("variantId")
        let productIndexParam = queryValue("productIndex")
        let quantityParam = queryValue("quantity") ?? "1"

        guard let quantity = Int(quantityParam), quantity > 0 else {
            throw CartBootstrapError.invalidQuantity
        }

        if variantId?.isEmpty == false, productIndexParam?.isEmpty == false {
            throw CartBootstrapError.ambiguousVariant
        }

        if let variantId, !variantId.isEmpty {
            return CartBootstrapLink(variantId: variantId, productIndex: nil, quantity: quantity)
        }

        guard let productIndexParam, !productIndexParam.isEmpty else {
            throw CartBootstrapError.missingVariant
        }

        guard let productIndex = Int(productIndexParam), productIndex >= 0 else {
            throw CartBootstrapError.invalidProductIndex
        }

        return CartBootstrapLink(variantId: nil, productIndex: productIndex, quantity: quantity)
    }
}

enum CartBootstrapError: LocalizedError {
    case unsupportedPath
    case missingVariant
    case ambiguousVariant
    case invalidQuantity
    case invalidProductIndex
    case productVariantNotFound

    var errorDescription: String? {
        switch self {
        case .unsupportedPath:
            return "Unsupported cart bootstrap path"
        case .missingVariant:
            return "Missing variantId or productIndex"
        case .ambiguousVariant:
            return "Use variantId or productIndex, not both"
        case .invalidQuantity:
            return "quantity must be a positive integer"
        case .invalidProductIndex:
            return "productIndex must be a non-negative integer"
        case .productVariantNotFound:
            return "Cart bootstrap product variant was not found"
        }
    }
}
